sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/FieldBuilder.java

```diff
@@ -0,0 +1,256 @@
+    }
+
+    private static Field buildField(java.lang.reflect.Field classField, Stack<Class<?>> classChain,
+        ClientLogger logger) {
+        Type type = classField.getGenericType();
+
+        if (SUPPORTED_NONE_PARAMETERIZED_TYPE.containsKey(type)) {
+            return buildNoneParameterizedType(classField, logger);
+        }
+        if (isArrayOrList(type)) {
+            return buildCollectionField(classField, classChain, logger);
+        }
+        List<Field> childFields = build((Class<?>) type, classChain, logger);
+        Field searchField = convertToBasicSearchField(classField, logger);
+        searchField.setFields(childFields);
+        return searchField;
+    }
+
+    private static Field buildNoneParameterizedType(java.lang.reflect.Field classField,
+        ClientLogger logger) {
+        Field searchField = convertToBasicSearchField(classField, logger);
+        return enrichWithAnnotation(searchField, classField, logger);
+    }
+
+
+    private static boolean isArrayOrList(Type type) {
+        return type.getClass().isArray() || isList(type);
+    }
+
+    private static boolean isList(Type type) {
+        if (!(type instanceof ParameterizedType)) {
+            return false;
+        }
+
+        Type rawType = ((ParameterizedType) type).getRawType();
+        return List.class.isAssignableFrom((Class<?>) rawType);
+    }
+
+    private static Field buildCollectionField(java.lang.reflect.Field classField,
+        Stack<Class<?>> classChain, ClientLogger logger) {
+        Type componentOrElementType = getComponentOrElementType(classField.getGenericType(), logger);
+        validateType(componentOrElementType, true, logger);
+        if (SUPPORTED_NONE_PARAMETERIZED_TYPE.containsKey(componentOrElementType)) {
+            Field searchField = convertToBasicSearchField(classField, logger);
+            return enrichWithAnnotation(searchField, classField, logger);
+        }
+        List<Field> childFields = build((Class<?>) componentOrElementType, classChain, logger);
+        Field searchField = convertToBasicSearchField(classField, logger);
+        searchField.setFields(childFields);
+        return searchField;
+    }
+
+    private static Type getComponentOrElementType(Type arrayOrListType, ClientLogger logger) {
+        if (arrayOrListType.getClass().isArray()) {
+            return arrayOrListType.getClass().getComponentType();
+        }
+        if (isList(arrayOrListType)) {
+            ParameterizedType pt = (ParameterizedType) arrayOrListType;
+            return pt.getActualTypeArguments()[0];
+        }
+        throw logger.logExceptionAsError(new RuntimeException(String.format(
+            "Collection type %s is not supported.", arrayOrListType.getTypeName())));
+    }
+
+    private static Field convertToBasicSearchField(java.lang.reflect.Field classField,
+        ClientLogger logger) {
+        Field searchField = new Field();
+        searchField.setName(classField.getName());
```

Comment:
The field name needs to be overridable. Imagine a key named "hid" but the field is `hotelId`.

Reply:
Same one as below.